File: test_generate_UnityToon.cs
using System;
using System.IO;
using System.Text;
using System.Text.RegularExpressions;

class Program
{
    static void Main()
    {
        try
        {
            // Test reading the common properties shader
            string commonPropertiesPath = "com.unity.toonshader/Runtime/Shaders/Common/Parts/CommonProperties.shaderblock";
            string commonPropertiesShader = File.ReadAllText(commonPropertiesPath);

            if (string.IsNullOrEmpty(commonPropertiesShader))
            {
                Console.WriteLine("ERROR: Common properties shader file is empty or could not be read");
                return;
            }
            string commonProperties = ExtractProperties(commonPropertiesShader);
            if (string.IsNullOrEmpty(commonProperties))
            {
                Console.WriteLine("ERROR: Failed to extract properties from CommonProperties.shaderblock");
                return;
            }

            Console.WriteLine($"Successfully extracted common properties. Length: {commonProperties.Length} characters");

            // Test reading the tessellation properties shader
            string tessellationPropertiesPath = "com.unity.toonshader/Runtime/Shaders/Common/Parts/TessellationProperties.shaderblock";
            string tessellationPropertiesShader = File.ReadAllText(tessellationPropertiesPath);

            if (string.IsNullOrEmpty(tessellationPropertiesShader))
            {
                Console.WriteLine("ERROR: Tessellation properties shader file is empty or could not be read");
                return;
            }
            string tessellationProperties = ExtractProperties(tessellationPropertiesShader);
            if (string.IsNullOrEmpty(tessellationProperties))
            {
                Console.WriteLine("ERROR: Failed to extract properties from TessellationProperties.shaderblock");
                return;
            }

            Console.WriteLine($"Successfully extracted tessellation properties. Length: {tessellationProperties.Length} characters");

            // Test reading the original shader files
            string unityToonPath = "com.unity.toonshader/Runtime/Integrated/Shaders/UnityToon.shader";
            string unityToonContent = File.ReadAllText(unityToonPath);

            if (string.IsNullOrEmpty(unityToonContent))
            {
                Console.WriteLine("ERROR: UnityToon.shader file is empty or could not be read");
                return;
            }

            Console.WriteLine($"Successfully read UnityToon.shader. Length: {unityToonContent.Length} characters");

            // Test the Properties block replacement
            string propertiesPattern = @"Properties\s*\{";
            Match startMatch = Regex.Match(unityToonContent, propertiesPattern);

            if (!startMatch.Success)
            {
                Console.WriteLine("ERROR: Could not find Properties block start in shader file");
                return;
            }

            Console.WriteLine($"Found Properties block start at position {startMatch.Index}");

            // Find the matching closing brace
            int startIndex = startMatch.Index;
            int braceCount = 0;
            int endIndex = startIndex;
            bool foundStart = false;

            for (int i = startIndex; i < unityToonContent.Length; i++)
            {
                if (unityToonContent[i] == '{')
                {
                    braceCount++;
                    foundStart = true;
                }
                else if (unityToonContent[i] == '}')
                {
                    braceCount--;
                    if (foundStart && braceCount == 0)
                    {
                        endIndex = i;
                        break;
                    }
                }
            }

            if (braceCount != 0)
            {
                Console.WriteLine("ERROR: Could not find matching closing brace for Properties block");
                return;
            }

            Console.WriteLine($"Found Properties block end at position {endIndex}");

            // Build new Properties block
            StringBuilder newProperties = new StringBuilder();
            newProperties.AppendLine("    Properties {");

            // Add common properties
            string[] commonLines = commonProperties.Split('\n');
            int propertyCount = 0;
            foreach (string line in commonLines)
            {
                if (!string.IsNullOrWhiteSpace(line) && !line.TrimStart().StartsWith("//"))
                {
                    newProperties.AppendLine($"        {line.Trim()}");
                    propertyCount++;
                }
            }

            newProperties.AppendLine("    }");

            Console.WriteLine($"Generated new Properties block with {propertyCount} properties. Length: {newProperties.Length} characters");

            // Test the replacement
            string newContent = unityToonContent.Substring(0, startIndex) + newProperties.ToString() + unityToonContent.Substring(endIndex + 1);
            newContent = ApplyAutoGeneratedComment(newContent, "//Auto-generated (test output)");

            Console.WriteLine($"Generated new shader content. Original length: {unityToonContent.Length}, New length: {newContent.Length}");

            // Write test file
            string testPath = "UnityToon_Generated_Test.shader";
            File.WriteAllText(testPath, newContent);

            Console.WriteLine($"Test shader written to {testPath}");
            Console.WriteLine("Shader generation test completed successfully!");

            try
            {
                File.Delete(testPath);
            }
            catch (IOException)
            {
            }
            catch (UnauthorizedAccessException)
            {
            }
        }
        catch (Exception e)
        {
            Console.WriteLine($"ERROR: {e.Message}");
            Console.WriteLine(e.StackTrace);
        }
    }

    private static string ExtractProperties(string shaderContent)
    {
        if (string.IsNullOrEmpty(shaderContent))
        {
            return null;
        }

        string propertiesPattern = @"Properties\s*\{";
        Match startMatch = Regex.Match(shaderContent, propertiesPattern);
        if (!startMatch.Success)
        {
            return null;
        }

        int openBraceIndex = shaderContent.IndexOf('{', startMatch.Index);
        if (openBraceIndex == -1)
        {
            return null;
        }

        int braceCount = 1;
        int closeBraceIndex = -1;

        for (int i = openBraceIndex + 1; i < shaderContent.Length; i++)
        {
            char c = shaderContent[i];
            if (c == '{')
            {
                braceCount++;
            }
            else if (c == '}')
            {
                braceCount--;
                if (braceCount == 0)
                {
                    closeBraceIndex = i;
                    break;
                }
            }
        }

        if (closeBraceIndex == -1)
        {
            return null;
        }

        string block = shaderContent.Substring(openBraceIndex + 1, closeBraceIndex - openBraceIndex - 1);
        string[] rawLines = block.Split(new[] { "\r\n", "\n", "\r" }, StringSplitOptions.None);

        for (int i = 0; i < rawLines.Length; i++)
        {
            rawLines[i] = rawLines[i].Trim();
        }

        int start = 0;
        int end = rawLines.Length - 1;

        while (start <= end && string.IsNullOrEmpty(rawLines[start]))
        {
            start++;
        }

        while (end >= start && string.IsNullOrEmpty(rawLines[end]))
        {
            end--;
        }

        if (start > end)
        {
            return string.Empty;
        }

        StringBuilder result = new StringBuilder();
        for (int i = start; i <= end; i++)
        {
            result.Append(rawLines[i]);
            if (i < end)
            {
                result.Append('\n');
            }
        }

        return result.ToString();
    }

    private static string ApplyAutoGeneratedComment(string content, string commentLine)
    {
        const string autoPrefix = "//Auto-generated on ";
        string[] lines = content.Split(new[] { "\n" }, StringSplitOptions.None);

        if (lines.Length > 0 && lines[0].StartsWith(autoPrefix, StringComparison.Ordinal))
        {
            lines[0] = commentLine;
        }
        else
        {
            string[] updatedLines = new string[lines.Length + 1];
            updatedLines[0] = commentLine;
            Array.Copy(lines, 0, updatedLines, 1, lines.Length);
            lines = updatedLines;
        }

        string result = string.Join("\n", lines);
        if (!result.EndsWith("\n", StringComparison.Ordinal))
        {
            result += "\n";
        }

        return result;
    }
}

File: test_generate_UnityToon.py
#!/usr/bin/env python3

import os
import re

from generate_UnityToon import (
    extract_properties_from_shader_content,
    apply_auto_generated_comment,
)


def test_shader_generation():
    try:
        # Test reading the common properties shader
        common_properties_path = "com.unity.toonshader/Runtime/Shaders/Common/Parts/CommonProperties.shaderblock"
        with open(common_properties_path, 'r') as f:
            common_properties_shader = f.read()

        if not common_properties_shader:
            print("ERROR: Common properties shader file is empty or could not be read")
            return

        common_properties = extract_properties_from_shader_content(common_properties_shader)
        if not common_properties:
            print("ERROR: Failed to extract common properties from shader")
            return

        print(f"Successfully extracted common properties. Length: {len(common_properties)} characters")

        # Test reading the tessellation properties shader
        tessellation_properties_path = "com.unity.toonshader/Runtime/Shaders/Common/Parts/TessellationProperties.shaderblock"
        with open(tessellation_properties_path, 'r') as f:
            tessellation_properties_shader = f.read()

        if not tessellation_properties_shader:
            print("ERROR: Tessellation properties shader file is empty or could not be read")
            return

        tessellation_properties = extract_properties_from_shader_content(tessellation_properties_shader)
        if not tessellation_properties:
            print("ERROR: Failed to extract tessellation properties from shader")
            return

        print(f"Successfully extracted tessellation properties. Length: {len(tessellation_properties)} characters")

        # Test reading the original shader files
        unity_toon_path = "com.unity.toonshader/Runtime/Integrated/Shaders/UnityToon.shader"
        with open(unity_toon_path, 'r') as f:
            unity_toon_content = f.read()

        if not unity_toon_content:
            print("ERROR: UnityToon.shader file is empty or could not be read")
            return

        print(f"Successfully read UnityToon.shader. Length: {len(unity_toon_content)} characters")

        # Test the Properties block replacement
        properties_pattern = r"Properties\s*\{"
        start_match = re.search(properties_pattern, unity_toon_content)

        if not start_match:
            print("ERROR: Could not find Properties block start in shader file")
            return

        print(f"Found Properties block start at position {start_match.start()}")

        # Find the matching closing brace
        start_index = start_match.start()
        brace_count = 0
        end_index = start_index
        found_start = False

        for i in range(start_index, len(unity_toon_content)):
            if unity_toon_content[i] == '{':
                brace_count += 1
                found_start = True
            elif unity_toon_content[i] == '}':
                brace_count -= 1
                if found_start and brace_count == 0:
                    end_index = i
                    break

        if brace_count != 0:
            print("ERROR: Could not find matching closing brace for Properties block")
            return

        print(f"Found Properties block end at position {end_index}")

        # Build new Properties block
        new_properties = []
        new_properties.append("    Properties {")

        # Add common properties
        common_lines = common_properties.split('\n')
        property_count = 0
        for line in common_lines:
            if line.strip() and not line.strip().startswith("//"):
                new_properties.append(f"        {line.strip()}")
                property_count += 1

        new_properties.append("    }")

        new_properties_text = '\n'.join(new_properties)

        print(f"Generated new Properties block with {property_count} properties. Length: {len(new_properties_text)} characters")

        # Test the replacement
        new_content = unity_toon_content[:start_index] + new_properties_text + unity_toon_content[end_index + 1:]
        new_content = apply_auto_generated_comment(new_content, "//Auto-generated (test output)")

        print(f"Generated new shader content. Original length: {len(unity_toon_content)}, New length: {len(new_content)}")

        # Write test file
        test_path = "UnityToon_Generated_Test.shader"
        with open(test_path, 'w') as f:
            f.write(new_content)

        print(f"Test shader written to {test_path}")
        print("Shader generation test completed successfully!")

        try:
            import os
            os.remove(test_path)
        except OSError:
            pass

    except Exception as e:
        print(f"ERROR: {e}")
        import traceback
        traceback.print_exc()


if __name__ == "__main__":
    test_shader_generation()
